Possible data race in inmemory dispatcher

**Describe the bug**
As part of a failed unit test for (should be totally unrelated change):

```
2020-05-19T18:56:35.059Z	DEBUG	inmemorychannel/message_dispatcher_test.go:216	Test servers	{"transformations server": "http://127.0.0.1:39505", "transformations failure server": "http://127.0.0.1:36615", "receiver server": "http://127.0.0.1:43653", "dead letter server": "http://127.0.0.1:35895"}
2020-05-19T18:56:35.060Z	INFO	swappable/swappable_message_handler.go:93	Updating config (-old +new)	{"httpHandler": "swappable", "diff": "  multichannelfanout.Config{\n- \tChannelConfigs: nil,\n+ \tChannelConfigs: []multichannelfanout.ChannelConfig{\n+ \t\t{\n+ \t\t\tNamespace: \"default\",\n+ \t\t\tName:      \"channela\",\n+ \t\t\tHostName:  \"channela.svc\",\n+ \t\t\tFanoutConfig: fanout.Config{\n+ \t\t\t\tSubscriptions: []v1beta1.SubscriberSpec{\n+ \t\t\t\t\t{\n+ \t\t\t\t\t\tUID:           \"aaaa\",\n+ \t\t\t\t\t\tGeneration:    1,\n+ \t\t\t\t\t\tSubscriberURI: s\"http://127.0.0.1:39505\",\n+ \t\t\t\t\t\tReplyURI:      s\"http://127.0.0.1:38045\",\n+ \t\t\t\t\t},\n+ \t\t\t\t\t{\n+ \t\t\t\t\t\tUID:           \"cccc\",\n+ \t\t\t\t\t\tGeneration:    1,\n+ \t\t\t\t\t\tSubscriberURI: s\"http://127.0.0.1:36615\",\n+ \t\t\t\t\t\tReplyURI:      s\"http://127.0.0.1:38045\",\n+ \t\t\t\t\t\tDelivery:      &v1beta1.DeliverySpec{DeadLetterSink: &v1.Destination{URI: s\"http://127.0.0.1:35895\"}},\n+ \t\t\t\t\t},\n+ \t\t\t\t},\n+ \t\t\t},\n+ \t\t},\n+ \t\t{\n+ \t\t\tNamespace: \"default\",\n+ \t\t\tName:      \"channelb\",\n+ \t\t\tHostName:  \"channelb.svc\",\n+ \t\t\tFanoutConfig: fanout.Config{\n+ \t\t\t\tSubscriptions: []v1beta1.SubscriberSpec{{UID: \"bbbb\", Generation: 1, SubscriberURI: s\"http://127.0.0.1:43653\"}},\n+ \t\t\t},\n+ \t\t},\n+ \t},\n  }\n"}
    TestDispatcher_dispatch: message_dispatcher_test.go:148: listen tcp :39505: bind: address already in use
    TestDispatcher_dispatch: message_dispatcher_test.go:291: Expected 202, Have 200

==================
WARNING: DATA RACE
Write at 0x00c0002003b9 by goroutine 33:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:656 +0x4c
  testing.(*common).Fatalf()
      /usr/local/go/src/testing/testing.go:724 +0x9d
  knative.dev/eventing/pkg/inmemorychannel.TestDispatcher_dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/inmemorychannel/message_dispatcher_test.go:291 +0x241b
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:992 +0x1eb
Previous write at 0x00c0002003b9 by goroutine 38:
  testing.(*common).FailNow()
      /usr/local/go/src/testing/testing.go:656 +0x4c
  testing.(*common).Fatal()
      /usr/local/go/src/testing/testing.go:718 +0x89
  knative.dev/eventing/pkg/inmemorychannel.TestDispatcher_dispatch.func1()
      /home/prow/go/src/knative.dev/eventing/pkg/inmemorychannel/message_dispatcher_test.go:148 +0x9f
Goroutine 33 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1043 +0x660
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1285 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:992 +0x1eb
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1283 +0x527
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1200 +0x2ff
  main.main()
      _testmain.go:50 +0x223
Goroutine 38 (finished) created at:
  knative.dev/eventing/pkg/inmemorychannel.TestDispatcher_dispatch()
      /home/prow/go/src/knative.dev/eventing/pkg/inmemorychannel/message_dispatcher_test.go:146 +0x8fb
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:992 +0x1eb
==================
    TestDispatcher_dispatch: testing.go:906: race detected during execution of test
--- FAIL: TestDispatcher_dispatch (0.11s)
    : testing.go:906: race detected during execution of test
FAIL
FAIL	knative.dev/eventing/pkg/inmemorychannel	1.398s
?   	knative.dev/eventing/pkg/kncloudevents	[no test files]
?   	knative.dev/eventing/pkg/kncloudevents/testing	[no test files]
skipped 5954 lines unfold_more
?   	knative.dev/eventing/test/test_images/sequencestepper	[no test files]
?   	knative.dev/eventing/test/test_images/transformevents	[no test files]
FAIL
Finished run, return code is 1
XML report written to /logs/artifacts/junit_jbuCDaXc.xml
Step failed: default_unit_test_runner
===========================
==== UNIT TESTS FAILED ====
===========================
==== Tue May 19 11:57:51 PDT 2020
===========================
Step failed: run_unit_tests
+ EXIT_VALUE=1
+ set +o xtrace
```


 **Expected behavior**
Tests to pass

 **To Reproduce**
Not sure, yet, just jotting down what the logs say:

 **Knative release version**
HEAD

 **Additional context**
Add any other context about the problem here such as proposed priority

https://prow.knative.dev/view/gcs/knative-prow/pr-logs/pull/knative_eventing/3110/pull-knative-eventing-unit-tests/1262818492571193344

--- a/pkg/broker/ingress/ingress_handler_test.go
+++ b/pkg/broker/ingress/ingress_handler_test.go
@@ -237,7 +237,7 @@ func TestIngressHandler_Start(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
 		if err := handler.Start(ctx); err != nil {
-			t.Fatal(err)
+			t.Error(err)
 		}
 	}()
 	// Need time for the handler to start up. Wait.

--- a/pkg/inmemorychannel/message_dispatcher_test.go
+++ b/pkg/inmemorychannel/message_dispatcher_test.go
@@ -145,7 +145,7 @@ func TestDispatcher_dispatch(t *testing.T) {
 	// Start the dispatcher
 	go func() {
 		if err := dispatcher.Start(serverCtx); err != nil {
-			t.Fatal(err)
+			t.Error(err)
 		}
 	}()
 

--- a/pkg/mtbroker/ingress/ingress_handler_test.go
+++ b/pkg/mtbroker/ingress/ingress_handler_test.go
@@ -202,7 +202,7 @@ func TestIngressHandler_Start(t *testing.T) {
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
 		if err := handler.Start(ctx); err != nil {
-			t.Fatal(err)
+			t.Error(err)
 		}
 	}()
 	// Need time for the handler to start up. Wait.

